runtime/runtime/src/lib.rs

```diff
@@ -497,6 +503,12 @@ impl Runtime {
         result.gas_burnt = exec_fee;
         // Executing actions one by one
         for (action_index, action) in action_receipt.actions.iter().enumerate() {
+            let action_hash = create_hash_upgradable(
+                apply_state.current_protocol_version,
+                &receipt.receipt_id,
+                &apply_state.last_block_hash,
+                u64::max_value() - action_index as u64,
```

Comment:
This implicitly encodes the type by counting action index backwards and receipt index and data index forwards. I guess the intention here is to make sure action hash does not overlap with generated receipt id. Unfortunately the calls to `create_hash_upgradable` are scattered across the code base so to make sure this does not happen one needs to find all usages of `create_hash_upgradable` and examine the way they are called, to make sure they do not produce overlapping ids.

I suggest we have 3 different implementations of `create_hash_upgradable` all located in one place and called something like `create_action_hash`, `create_data_id`, `create_receipt_id_from_receipt`, `create_receipt_id_from_transaction` to make it easy to examine visually that they produce non-overlapping ids. These implementation can also take strongly-typed arguments, e.g. `fn create_receipt_id_from_transaction(protocol_version: ProtocolVersion, transaction: &SignedTransaction, last_block_hash: &CryptoHash)` to hide stuff like `u64::max_value() - action_index as u64` internally and make sure it is not misused.

Reply:
I agree on having a separation for `create_receipt_id` vs `create_random_seed`, etc.

But we still need to use salt/nonce to generate receipt IDs for new receipts.

Reply:
Ideally, different types of salt like ActionIndex, ReceiptIndex, etc need to be strongly typed (`struct ActionIndex(u64)`), to not be misused, and compiler would optimize it out anyway. However, let's not do it in this PR but prepare it for this change by separating `create_receipt_id`, etc. Me and @frol were discussing the next steps towards the runtime spec, and the major step can be introducing strong typeness into the runtime, to make the methods as self-descriptive as possible.

Reply:
Updated code to be explicit for every usage.